rasa/shared/core/training_data/story_reader/markdown_story_reader.py

```diff
@@ -231,6 +231,11 @@ def parse_e2e_message(line: Text) -> Message:
         intent = match.group(2)
         message = match.group(4)
         example = entities_parser.parse_training_example(message, intent)
+        if not unfold_utterances:
```

Comment:
Why is this a change in the `MarkdownReader`. I would expect that we have to change the `YAMLStoryWriter` to inline the entity annotations when `use_e2e=True`?

Reply:
Because the "unfoling" of i.g. "OR" statements happens during reading. The writer takes a list of StorySteps that are already modified 😬

Reply:
So the `unfold_utterances=False` is used only with `StoryMarkdownToYamlConverter`

Reply:
Sorry, I'm not getting this 🙈 Can't the writer "fold" it back?

Reply:
No, it can't

Reply:
"OR" -> copy over the steps that were before with no back trace.

Reply:
Ok, but why don't we parse the entities as we're used to in `parse_e2e_messages` and then fold them back in in `YAMLStoryWriter` whenever we write an end-to-end-test story?

Reply:
Because we replace the original text of the training example and remove all `{` `}` from it, so we don't have this information anymore.

Reply:
And this is exactly what this PR is targeting - preserving this information, bc for simple conversion we don't want to replace the original text.

Reply:
> Because we replace the original text of the training example and remove all { } from it, so we don't have this information anymore.

But although we don't have the entities in the text anymore, we still got is as `entities` properties within the `UserUttered` events, do we? So we can generate the text including the brackets in the writer.

Reply:
Just checked the `YAMLStoryWriter` code: so the `YAMLStoryWriter` has no clue whether it's currently in end-to-end mode or not? And we are trying to achieve different behavior in the output format by making the reader change?

Reply:
Yes, but why would we do that? 😄

Reply:
It seems super awkward to me to change the reader if we want to have a different output format. Also, we need this when dumping actual (not converting) end-to-end stories, don't we?

Reply:
@degiz another solution I could imagine is to subclass `MarkdownStoryReader` with something like `MarkdownStoryReaderForConversion` which implements the special behavior which you need for converting the stories. This way the converter-specific behavior would be isolated and clearly identifiable

Reply:
@wochinge  I like the idea with MarkdownStoryReaderForConversion, but that's quite an overkill for a reader that we deprecate, because that would require some refactorings.
Do you think it's clearer what's happening if I rename unfold_utterances to keep_original_user_utterances? Bc that's what it does essentially - keeps utterance the same as they were in a text file.

Reply:
how about naming it `is_used_for_conversion`? Then it's clear to everyone that this is a flag which should only be used when converting training data, but at no point else?

Reply:
sounds good to me 👍

Reply:
you have to invert your boolean logic then 🙈